Pass correct in-progress array pointer for writing to file

asimrp opened the pull request:
In case of extended queries, snapshot information is shared between reader and writer QEs using files.  The writer obtains a snapshot and writes it to file.  This patch fixes a bug that passed incorrect address of the array of in-progress transactions for writing to file.  The bug caused hard to reproduce
errors in production workloads that involved extended queries (bind/execute libpq messages, declare cursor statements, certain PL/* statements such as RETURN and EXECUTE in pl/pgsql).

To be back ported to all supported branches.

berlin-ab commented on src/backend/utils/time/test/sharedsnapshot_test.c:
```diff
@@ -5,9 +5,100 @@
+
+	/* create a dummy shared and local snapshot with 5 in-progress transactions */
+	SharedSnapshotSlot slot;
+	SharedLocalSnapshotSlot = &slot;
```
All of these values were necessary to set to trigger the bug?

Very cool that you did this with a test.

asimrp replied:
Yeah, I managed to iterate over and over until all the necessary state was created.  So yes, all these values are necessary.

Creating the `base/pgsql_tmp` directory was also an interesting exercise, it forced me to learn some GNU Make basics :-)

d commented on src/backend/utils/time/sharedsnapshot.c:
```diff
@@ -714,7 +714,7 @@ dumpSharedLocalSnapshot_forCursor(void)
 		FileWriteFieldWithCount(count, f, src->snapshot.xcnt);
 
-		if (!FileWriteOK(f, &src->snapshot.xip, src->snapshot.xcnt * sizeof(TransactionId)))
+		if (!FileWriteOK(f, src->snapshot.xip, src->snapshot.xcnt * sizeof(TransactionId)))
```
I spend a good 2 minutes thinking: how could this have _not_ blown up before? My best uneducated guess is:

The layout in shmem for the shared snapshot "slots" are pretty contiguous:

```
+-------------+-------------+-------------+------------------+-----------------------------+---------------------------+-----------------+-------------------------------+--------------------------+------------+-------------------------------+----------------------------+------------+----------------------+-------------------+-------------------+
|             |             |             |                  |                      Shared Snapshot Slot                                 |                               Shared Snapshot Slot                    |                              Shared Snapshot Slot                       |                      |                   |                   |
| numSlots(0) | maxSlots(4) | nextSlot(8) |   slots(0x0c)    | first part of a slot (0x14) | slot.snapshot.xip (0xdf4) | trailer (0xdfc) | first part of a slot (0x0e3c) | slot.snapshot.xip (1c1c) |  trailer   | first part of a slot (0x1c64) | slot.snapshot.xip (0x2a44) |  trailer   |   205 xids(0x2a8c)   | 205 xids (0x2dc0) | 205 xids (0x30f4) |
+-------------+-------------+-------------+------------------+-----------------------------+---------------------------+-----------------+-------------------------------+--------------------------+------------+-------------------------------+----------------------------+------------+----------------------+-------------------+-------------------+
|    00000005 |    00000005 |    00000004 | 0000cfcf000030f4 | <3552 bytes>                | 0000000000001c64          | <64 bytes>      | <3552 bytes>                  | 0000cfcf00002a8c         | <64 bytes> | <3552 bytes>                  | 0000cfcf00030f4            | <64 bytes> | 205 xids <820 bytes> | <820 bytes>       | <820 bytes>       |
+-------------+-------------+-------------+------------------+-----------------------------+---------------------------+-----------------+-------------------------------+--------------------------+------------+-------------------------------+----------------------------+------------+----------------------+-------------------+-------------------+
```

The most plausible "not blow-up" scenario is we use the ultimate slot, whose `xip` location (offset 0x2a44) is very close to the block of xids (offset 2a8c) used by the first snapshot slot. But even in that scenario, the initial 18 transaction ID's will be pretty bogus, right @asimrp ?

asimrp replied:
That's tight @d. All slots followed by xip arrays for all slots are a single contiguous chunk of memory.  Passing address of the xip member (instead of its content / the address it's pointing to) would never blow up because the (`xcnt * sizeof(TransactionId)`) bytes following it will always point to some region with the contiguous chunk.  In other words, we end up writing some window of `xcnt * sizeof(TransactionId)` bytes within this contiguous shared memory chunk to the file.  As there are no invalid (unallocated) addresses involved, things do not blow-up.

asimrp commented:
PR pipeline failed due to an unrelated failure in gpfdist_ssl test.  I've rebased against master and squashed the unittest fixes.  Hopefully the PR pipeline runs again and passes.

asimrp commented:
Pushed in spite of one red job on master CI because the one red job, gpexpand is being worked on and this fix is needed urgently on back branches as well as it is unrelated to the CI failure.

asimrp commented:
Merged to 6X_STABLE as 8eefffec605d6f8c0a86a0a8d903e67407643bc1 and 4fde3b6977d28315969db3b96eb43a4b92216155

asimrp commented:
Merged to 5X_STABLE as 2e719b1079a627f04a27267fc55a7c6f75921225, 906c0d0945fb106d6f234d22032139b824d60ea3 and ba2cb32fa45f259f1dadc3adeac3881c8009b25e
